src/libcore/slice.rs

```diff
@@ -684,13 +666,11 @@ macro_rules! iterator {
             fn next(&mut self) -> Option<$elem> {
                 // could be implemented with slices, but this avoids bounds checks
                 unsafe {
-                    ::intrinsics::assume(!self.ptr.is_null());
-                    ::intrinsics::assume(!self.end.is_null());
```

Comment:
They're potentially invalid if the element size is zero, because the iterators are free to wrap around in that case.

I also question the edge case of a slice of non-zero-sized elements that includes the highest byte in the address space. It should be theoretically possible to have a valid slice that includes an element placed in the absolute highest address, and if so, then the `end` pointer will have wrapped around to `0` and the `ptr` value will wrap around when the iterator yields its final element.